File: apps/android/world-inspector/src/main/java/com/localgpt/inspector/InspectorClient.kt
/// WebSocket client for connecting to the Bevy World Inspector server.

package com.localgpt.inspector

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

enum class ConnectionState {
    DISCONNECTED, CONNECTING, CONNECTED, ERROR
}

class InspectorClient {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val okhttp = OkHttpClient()
    private var webSocket: WebSocket? = null

    // State flows
    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _sceneTree = MutableStateFlow<List<TreeEntity>>(emptyList())
    val sceneTree: StateFlow<List<TreeEntity>> = _sceneTree.asStateFlow()

    private val _selectedEntityId = MutableStateFlow<ULong?>(null)
    val selectedEntityId: StateFlow<ULong?> = _selectedEntityId.asStateFlow()

    private val _selectedEntityDetail = MutableStateFlow<EntityDetailData?>(null)
    val selectedEntityDetail: StateFlow<EntityDetailData?> = _selectedEntityDetail.asStateFlow()

    private val _worldInfo = MutableStateFlow<WorldInfoData?>(null)
    val worldInfo: StateFlow<WorldInfoData?> = _worldInfo.asStateFlow()

    private val _selectedTransform = MutableStateFlow<Pair<List<Float>, List<Float>>?>(null)
    val selectedTransform: StateFlow<Pair<List<Float>, List<Float>>?> = _selectedTransform.asStateFlow()

    // Connection

    fun connect(host: String = "localhost", port: Int = 9877) {
        if (_connectionState.value == ConnectionState.CONNECTING ||
            _connectionState.value == ConnectionState.CONNECTED
        ) return

        _connectionState.value = ConnectionState.CONNECTING

        val request = Request.Builder()
            .url("ws://$host:$port/ws")
            .build()

        webSocket = okhttp.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                scope.launch { _connectionState.value = ConnectionState.CONNECTED }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                scope.launch { handleMessage(text) }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                scope.launch { _connectionState.value = ConnectionState.ERROR }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                scope.launch { _connectionState.value = ConnectionState.DISCONNECTED }
            }
        })
    }

    fun disconnect() {
        webSocket?.close(1000, "Client disconnect")
        webSocket = null
        _connectionState.value = ConnectionState.DISCONNECTED
        _sceneTree.value = emptyList()
        _selectedEntityId.value = null
        _selectedEntityDetail.value = null
        _worldInfo.value = null
        _selectedTransform.value = null
    }

    // Commands

    fun selectEntity(entityId: ULong) {
        send(ClientMessage.SelectEntity(entityId))
        _selectedEntityId.value = entityId
        _selectedEntityDetail.value = null
        send(ClientMessage.RequestEntityDetail(entityId))
    }

    fun deselect() {
        send(ClientMessage.Deselect())
        _selectedEntityId.value = null
        _selectedEntityDetail.value = null
        _selectedTransform.value = null
    }

    fun toggleVisibility(entityId: ULong) {
        send(ClientMessage.ToggleVisibility(entityId))
    }

    fun focusEntity(entityId: ULong) {
        send(ClientMessage.FocusEntity(entityId))
    }

    fun refreshSceneTree() {
        send(ClientMessage.RequestSceneTree())
    }

    fun refreshWorldInfo() {
        send(ClientMessage.RequestWorldInfo())
    }

    // Private

    private fun send(message: ClientMessage) {
        val json = InspectorJson.encodeToString(ClientMessage.serializer(), message)
        webSocket?.send(json)
    }

    private fun handleMessage(text: String) {
        val message = try {
            InspectorJson.decodeFromString(ServerMessage.serializer(), text)
        } catch (e: Exception) {
            return
        }

        when (message) {
            is ServerMessage.SceneTree -> {
                _sceneTree.value = message.entities
            }
            is ServerMessage.EntityDetail -> {
                if (_selectedEntityId.value == message.entityId) {
                    _selectedEntityDetail.value = message.data
                }
            }
            is ServerMessage.WorldInfo -> {
                _worldInfo.value = message.data
            }
            is ServerMessage.SelectionChanged -> {
                _selectedEntityId.value = message.entityId
                _selectedEntityDetail.value = null
                _selectedTransform.value = null
                send(ClientMessage.RequestEntityDetail(message.entityId))
            }
            is ServerMessage.SelectionCleared -> {
                _selectedEntityId.value = null
                _selectedEntityDetail.value = null
                _selectedTransform.value = null
            }
            is ServerMessage.SceneChanged -> {
                send(ClientMessage.RequestSceneTree())
            }
            is ServerMessage.EntityTransformUpdated -> {
                if (_selectedEntityId.value == message.entityId) {
                    _selectedTransform.value = Pair(message.position, message.rotationDegrees)
                }
            }
        }
    }
}

File: apps/apple/WorldInspector/Sources/WorldInspector/InspectorClient.swift
/// WebSocket client for connecting to the Bevy World Inspector server.

import Foundation
import Observation

/// Connection state for the inspector client.
public enum ConnectionState: Equatable {
    case disconnected
    case connecting
    case connected
    case error(String)
}

/// Observable client that manages the WebSocket connection and world state.
@Observable
public final class InspectorClient {
    // MARK: - Published State

    /// Current connection state.
    public private(set) var connectionState: ConnectionState = .disconnected

    /// The entity tree from the server.
    public private(set) var sceneTree: [TreeEntity] = []

    /// Currently selected entity ID.
    public private(set) var selectedEntityId: UInt64?

    /// Detail data for the selected entity.
    public private(set) var selectedEntityDetail: EntityDetailData?

    /// Global world info.
    public private(set) var worldInfo: WorldInfoData?

    /// Live transform of the selected entity (streamed at 10Hz).
    public private(set) var selectedTransform: (position: SIMD3<Float>, rotation: SIMD3<Float>)?

    // MARK: - Private

    private var webSocketTask: URLSessionWebSocketTask?
    private let session = URLSession(configuration: .default)
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    public init() {}

    // MARK: - Connection

    /// Connect to the inspector WebSocket server.
    public func connect(host: String = "localhost", port: Int = 9877) {
        guard connectionState != .connecting && connectionState != .connected else { return }

        connectionState = .connecting
        let url = URL(string: "ws://\(host):\(port)/ws")!
        webSocketTask = session.webSocketTask(with: url)
        webSocketTask?.resume()
        connectionState = .connected
        receiveMessages()
    }

    /// Disconnect from the server.
    public func disconnect() {
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
        webSocketTask = nil
        connectionState = .disconnected
        sceneTree = []
        selectedEntityId = nil
        selectedEntityDetail = nil
        worldInfo = nil
        selectedTransform = nil
    }

    // MARK: - Commands

    /// Select an entity by ID. Syncs across all connected clients.
    public func selectEntity(_ entityId: UInt64) {
        send(.selectEntity(entityId: entityId))
        selectedEntityId = entityId
        selectedEntityDetail = nil
        send(.requestEntityDetail(entityId: entityId))
    }

    /// Clear the selection.
    public func deselect() {
        send(.deselect)
        selectedEntityId = nil
        selectedEntityDetail = nil
        selectedTransform = nil
    }

    /// Toggle visibility of an entity.
    public func toggleVisibility(_ entityId: UInt64) {
        send(.toggleVisibility(entityId: entityId))
    }

    /// Focus the Bevy camera on an entity.
    public func focusEntity(_ entityId: UInt64) {
        send(.focusEntity(entityId: entityId))
    }

    /// Request a fresh scene tree.
    public func refreshSceneTree() {
        send(.requestSceneTree)
    }

    /// Request fresh world info.
    public func refreshWorldInfo() {
        send(.requestWorldInfo)
    }

    // MARK: - Private

    private func send(_ message: ClientMessage) {
        guard let data = try? encoder.encode(message),
              let text = String(data: data, encoding: .utf8)
        else { return }

        webSocketTask?.send(.string(text)) { [weak self] error in
            if let error {
                self?.connectionState = .error(error.localizedDescription)
            }
        }
    }

    private func receiveMessages() {
        webSocketTask?.receive { [weak self] result in
            guard let self else { return }

            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self.handleTextMessage(text)
                case .data:
                    break // Binary frames (GLB snapshots) — not yet handled
                @unknown default:
                    break
                }
                // Continue receiving
                self.receiveMessages()

            case .failure(let error):
                self.connectionState = .error(error.localizedDescription)
            }
        }
    }

    private func handleTextMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let message = try? decoder.decode(ServerMessage.self, from: data)
        else { return }

        Task { @MainActor in
            switch message {
            case .sceneTree(let entities):
                self.sceneTree = entities

            case .entityDetail(let entityId, let detail):
                if self.selectedEntityId == entityId {
                    self.selectedEntityDetail = detail
                }

            case .worldInfo(let info):
                self.worldInfo = info

            case .selectionChanged(let entityId):
                self.selectedEntityId = entityId
                self.selectedEntityDetail = nil
                self.selectedTransform = nil
                self.send(.requestEntityDetail(entityId: entityId))

            case .selectionCleared:
                self.selectedEntityId = nil
                self.selectedEntityDetail = nil
                self.selectedTransform = nil

            case .sceneChanged:
                self.send(.requestSceneTree)

            case .entityTransformUpdated(let entityId, let pos, let rot):
                if self.selectedEntityId == entityId {
                    self.selectedTransform = (position: pos, rotation: rot)
                }
            }
        }
    }
}
